File: bin/random_color.py
#!/usr/bin/env python3
''' A thin wrapper around randomcolor.RandomColor.generate() '''

## IMPORTS
from argparse import ArgumentParser
from randomcolor import RandomColor

# Define a function that drops None type arguments.
def rmNone(args):
    rm = [key for key in args if args.get(key) is None]
    for key in rm: del args[key]
    return(args)


## INPUTS
ap = ArgumentParser(description='Generate a random color.')

# Count
ap.add_argument('-c','--count', default = 1, type = int, help='''
An integer which specifies the number of colors to generate.''')

# Hue
ap.add_argument('-u','--hue', type = str, default=None, help='''
Controls the hue of the generated color. You can pass a string
representing a color name: red, orange, yellow, green, blue, purple, pink and
monochrome are currently supported.''')

# Luminosity
ap.add_argument('-l','--luminosity', type = str, default=None, help='''
Controls the luminosity of the generated color. You can specify a
string containing bright, light, or dark.''')

# Seed
# FIXME: doesnt work
#ap.add_argument('-s','--seed', type = int, default=None, help='''
#An integer which when passed will cause randomColor to return
#the same color each time.''')

# Format
ap.add_argument('-f','--format', type = str, default=None, help='''
A string which specifies the format of the generated color. Possible
values are rgb, rgbArray, hsl, hslArray and hex (default).''')


# Parse input arguments.
args = vars(ap.parse_args())
args = rmNone(args)

# Generate n random colors.
rand_color = RandomColor()
print(rand_color.generate(**args))

File: R/random_color.R
#' random_color
#'
#' @export

random_color <-  function(count=1,hue=NULL,luminosity="light"){
       # Count - number of colors to generate.
       # Hue - a string 
       # luminosity - a string, one of c(light,dark,bright).
	
	# Path to python script.
	root <- getrd()
	script <- file.path(root,"Py","random_color.py")

	# Collect arguments in a list.
	args_list <- list("count"=count,"hue"=hue,
			  "luminosity"=luminosity)

	# Drop NULL args.
	drop <- which(sapply(args_list,is.null))
	if (length(drop) > 0) {
		args_list <- args_list[-drop]
	}

	# Format args as options passed to python script.
	options_ <- paste(paste(paste0("--",names(args_list)),args_list),
			  collapse=" ")

	# Evaluate call to random_color.py
	cmd <- paste(script, options_)
	response <- system(cmd,intern=TRUE)

	# Return the hex color(s).
	hex <- toupper(rjson::fromJSON(gsub("'","\"",response)))

	return(hex)
}
